backend/pkg/database/refresh_materialized_views.go

```diff
@@ -0,0 +1,81 @@
+// Copyright © 2022 Cisco Systems, Inc. and its affiliates.
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+const materializedViewRefreshIntervalSecond = 5
+
+type ViewRefreshHandler struct {
+	mu                 sync.Mutex
+	shouldRefreshViews bool
+}
+
+func (vh *ViewRefreshHandler) SetTrue() {
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	vh.shouldRefreshViews = true
+}
```

Comment:
If we let this function accept a string and we added `tablesChanged map[string]bool` to the ViewRefreshHandler we could create a set of which tables have changed, with that information we could change the view handler to allow views to register their own refresh handlers based on different table changes. I'm thinking something like:

```
type refreshfunc func(db *gorm.DB)

type ViewRefreshHandler struct {
	mu                        sync.Mutex
	tablesChanged   map[string]bool
	refreshHandlers map[string][]refreshfunc
}

func (vh *ViewRefreshHandler) TableChanged(table string) {
    	vh.mu.Lock()
	defer vh.mu.Unlock()
	vh.tablesChanged[table] = true
}

func (vh *ViewRefreshHandler) GetAndClearChanges() map[string]bool {
	vh.mu.Lock()
	defer vh.mu.Unlock()
	changes := vh.tableChanged
	vh.tableChanged = map[string]bool{}
	return changes
}

func (vh *ViewRefreshHandler) RegisterViewRefreshHandler(f refreshfunc, tables ...string) {
        for _, table := range tables {
            vh.refreshHandlers[table] = append(vh.RefreshHandlers[table], f)
        }
}

func (vh *ViewRefreshHandler) DoRequiredRefreshes(db *gorm.DB) {
	changedTables = vh.GetAndClearChanges()
         for table, _ := range changedTables {
              handlers := vh.refreshHandlers[table]
              for _, handler := range handlers {
                  handler(db)
              }
         }
}

func (db *Handler) RefreshMaterializedViews() {
	ticker := time.NewTicker(materializedViewRefreshIntervalSecond * time.Second)
	
	done := make(chan bool)
	for {
		select {
		case <-done:
			return
		case <-ticker.C:
                        db.ViewRefreshHandler.DoRequiredRefreshes(db)
		}
	}
}
```

Reply:
Thanks for the suggestion, I will do it this way.